pkg/config/apiserver.go

```diff
@@ -0,0 +1,235 @@
+type APIResponse struct {
+	Data   interface{} `json:"data,omitempty"`
+	Status *rpc.Status `json:"status,omitempty"`
+}
+
+// register routes
+func (a *API) register(c *restful.Container) {
+	ws := &restful.WebService{}
+	ws.Consumes(restful.MIME_JSON, "application/yaml", "application/x-yaml")
+	ws.Produces(restful.MIME_JSON)
+	ws.Path(a.rootPath)
+
+	ws.Route(ws.
+		GET("/scopes/{scope}/subjects/{subject}/rules").
+		To(a.getRules).
+		Doc("Gets rules associated with the given scope and subject").
+		Param(ws.PathParameter("scope", "scope").DataType("string")).
+		Param(ws.PathParameter("subject", "subject").DataType("string")).
+		Writes(APIResponse{}))
+
+	ws.Route(ws.
+		PUT("/scopes/{scope}/subjects/{subject}/rules").
+		To(a.putRules).
+		Doc("Replaces rules associated with the given scope and subject").
+		Param(ws.PathParameter("scope", "scope").DataType("string")).
+		Param(ws.PathParameter("subject", "subject").DataType("string")).
+		Writes(APIResponse{}))
+	c.Add(ws)
+}
+
+// NewAPI creates a new API server
+func NewAPI(version string, port uint16, eval expr.Validator, aspectFinder AspectValidatorFinder,
+	builderFinder BuilderValidatorFinder, findAspects AdapterToAspectMapper, store KeyValueStore) *API {
+	c := restful.NewContainer()
+	a := &API{
+		version:  version,
+		rootPath: fmt.Sprintf("/api/%s", version),
+		store:    store,
+		readBody: ioutil.ReadAll,
+		validate: func(cfg map[string]string) (*Validated, descriptor.Finder, *adapter.ConfigErrors) {
+			v := newValidator(aspectFinder, builderFinder, findAspects, true, eval)
+			rt, ce := v.validate(cfg)
+			return rt, v.descriptorFinder, ce
+		},
+	}
+	a.register(c)
+	a.server = &http.Server{Addr: ":" + strconv.Itoa(int(port)), Handler: c}
+	a.handler = c
+	// ensure that we always send back an APIResponse object.
+	c.ServiceErrorHandler(func(err restful.ServiceError, req *restful.Request, resp *restful.Response) {
+		writeErrorResponse(err.Code, err.Message, resp)
+	})
+	return a
+}
+
+// Run calls listen and serve on the API server
+func (a *API) Run() {
+	glog.Warning(a.server.ListenAndServe())
+}
+
+// getRules returns the entire service config document for the scope and subject
+// "/scopes/{scope}/subjects/{subject}/rules"
+func (a *API) getRules(req *restful.Request, resp *restful.Response) {
+	funcPath := req.Request.URL.Path[len(a.rootPath):]
+	st, msg, data := getRules(a.store, funcPath)
+
+	writeResponse(st, msg, data, resp)
+
+	/*
+		// TODO send index back to the client
+		//_ = index
+	*/
+}
+
+func getRules(store KeyValueStore, path string) (statusCode int, msg string, data *pb.ServiceConfig) {
+	var val string
+	var found bool
+
+	if val, _, found = store.Get(path); !found {
+		return http.StatusNotFound, fmt.Sprintf("no rules for %s", path), nil
+	}
+
+	m := &pb.ServiceConfig{}
+	if err := yaml.Unmarshal([]byte(val), m); err != nil {
+		return http.StatusInternalServerError, fmt.Sprintf("unbale to parse rules at %s", path), nil
```

Comment:
Please include the text for 'err' in the final message, or log it or something. As it is, it's being lost.